File: Lista/quest04/heranca.go
package main

import "fmt"

// Interface que define o comportamento comum
type Animal interface {
	Som() string
}

// Estrutura base
type Cachorro struct{}

func (c Cachorro) Som() string {
	return "Au Au"
}

type Gato struct{}

func (g Gato) Som() string {
	return "Miau"
}

func main() {
	var animal Animal

	animal = Cachorro{}
	fmt.Println("Cachorro:", animal.Som())

	animal = Gato{}
	fmt.Println("Gato:", animal.Som())
}

File: Lista/quest04/heranca.java
abstract class Animal {
    public abstract void som();
}

class Cachorro extends Animal {
    @Override
    public void som() {
        System.out.println("Au Au");
    }
}

class Gato extends Animal {
    @Override
    public void som() {
        System.out.println("Miau");
    }
}

public class Main {
    public static void main(String[] args) {
        Animal cachorro = new Cachorro();
        cachorro.som();

        Animal gato = new Gato();
        gato.som();
    }
}
